doc/compiler/il/notes_mainline_outOfLine_GlRegDeps.md

```diff
@@ -0,0 +1,97 @@
+<!--
+Copyright (c) 2021, 2021 IBM Corp. and others
+
+This program and the accompanying materials are made available under
+the terms of the Eclipse Public License 2.0 which accompanies this
+distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+or the Apache License, Version 2.0 which accompanies this distribution and
+is available at https://www.apache.org/licenses/LICENSE-2.0.
+
+This Source Code may also be made available under the following
+Secondary Licenses when the conditions for such availability set
+forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+General Public License, version 2 with the GNU Classpath
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# Mainline and out of line code
+The mainline path handles the common operations in a straight line to avoid branching or using branch prediction resources. Arranging the commonly run code on the mainline also helps reserve instruction cache. The work can be done in the same cache line by falling through. There is no need to jump to another spot to bring in an entire new cache line to the cache.
+
+Executing the common code on the mainline is not only adopted in IL but also in the code gen, such as the non-trivial logics: write barrier, instanceof, etc. The instructions that are believed to run commonly are arranged on the mainline path. Branching is chosen for the less common code.
+
+One example in the code gen is the bound check. It does the comparison between arraylength and index. If it fails, an exception is going to be raised.  Then it will branch to an out of line, or a snippet of code that jumps into the VM to throw the exception that needs to be thrown. Because the path that jumps to throw the exception is expected to be very rarely called, the snippet code is put at the end of the code section, or at the end of the generated code for that method, aka out of line, so that it does not interfere with the code that we believe will actually run.
+
+Another example is block ordering and extension of basic blocks in IL. Both uses the same logic to arrange the blocks. For example there is a block ending with `if`. One path is marked as cold. Another path is not marked as cold. Block ordering takes the cold path out of the mainline, puts it at the end of the method, and keeps the non-cold path as the fall through. This way it avoids consuming branch prediction and instruction cache resources.
+
```

Comment:
"For example..." sentence does not make it completely clear that the one path marked cold and the other not marked as cold is just one possible scenario.